### PR TITLE
[core] Fixes for upcoming eslint upgrade

### PR DESCRIPTION
**Pull request by Janpot**
Bringing in most of the fixes necessary to pass the eslint upgrade in https://github.com/mui/material-ui/pull/34642

Nothing should be breaking. Just pointing out changes to the following: https://github.com/mui/mui-x/pull/6667/files#diff-8a3d1997d458bbb8861a31ed9b2db9137c1389883964b6d546c68b42645f5c96

### CODE REVIEW

File: benchmark/browser/scripts/benchmark.js [hunk 1]
```diff
@@ -9,7 +9,9 @@ const http = require('http');
 const PORT = 1122;
 
 function sleep(ms) {
-  return new Promise((resolve) => setTimeout(resolve, ms));
+  return new Promise((resolve) => {
+    setTimeout(resolve, ms);
+  });
```
Comment (oliviertassinari) [inline benchmark/browser/scripts/benchmark.js:14]: Nice, I have always hated these implicit unclear returns.

File: packages/grid/x-data-grid-premium/src/hooks/features/rowGrouping/useGridRowGrouping.tsx [hunk 2]
```diff
@@ -258,7 +260,7 @@ export const useGridRowGrouping = (
   >(() => {
     const sanitizedRowGroupingModel = gridRowGroupingSanitizedModelSelector(apiRef);
     const rulesOnLastRowTreeCreation =
-      apiRef.current.unstable_caches.rowGrouping.rulesOnLastRowTreeCreation;
+      apiRef.current.unstable_caches.rowGrouping.rulesOnLastRowTreeCreation || [];
```
Comment (oliviertassinari) [inline packages/grid/x-data-grid-premium/src/hooks/features/rowGrouping/useGridRowGrouping.tsx:263]: Should we have behavior changes in this PR?

```suggestion
      apiRef.current.unstable_caches.rowGrouping.rulesOnLastRowTreeCreation;
```

File: packages/x-date-pickers/src/ClockPicker/ClockPicker.tsx [hunk 3]
```diff
@@ -159,7 +159,7 @@ export const ClockPicker = React.forwardRef(function ClockPicker<TDate extends u
     components,
     componentsProps,
     value,
-    disableIgnoringDatePartForTimeValidation,
+    disableIgnoringDatePartForTimeValidation = false,
```
Comment (oliviertassinari) [inline packages/x-date-pickers/src/ClockPicker/ClockPicker.tsx:162]:
```suggestion
    disableIgnoringDatePartForTimeValidation,
```

**Review by oliviertassinari — COMMENTED** [review]
👍

File: packages/grid/x-data-grid-premium/src/hooks/features/rowGrouping/useGridRowGrouping.tsx (hunk 2 above)
Comment (Janpot) [reply packages/grid/x-data-grid-premium/src/hooks/features/rowGrouping/useGridRowGrouping.tsx:263]: It replaces https://github.com/mui/mui-x/pull/6667/files#diff-1250ff9bde7c6785848b32e7acde51bdce6c5e7d4d82a884a287d33e0e4d8350L267 which eslint complains in `https://eslint.org/docs/latest/rules/default-param-last`.

I can ignore it as well

File: packages/x-date-pickers/src/ClockPicker/ClockPicker.tsx (hunk 3 above)
Comment (Janpot) [reply packages/x-date-pickers/src/ClockPicker/ClockPicker.tsx:162]: Same as above but for https://github.com/mui/mui-x/pull/6667/files#diff-e2061e7eec043ba1ab45746803975267576725a37d32733b469fc7f8b2079a00L40.

Happy to revert and ignore, but this felt quite safe as it didn't seem to be publicly exposed.

File: packages/grid/x-data-grid-premium/src/hooks/features/rowGrouping/useGridRowGrouping.tsx (hunk 2 above)
Comment (Janpot) [reply packages/grid/x-data-grid-premium/src/hooks/features/rowGrouping/useGridRowGrouping.tsx]: As far as I can tell it's functionally equivalent, but I removed it. We can bring it back in a separate PR.

File: packages/x-date-pickers/src/ClockPicker/ClockPicker.tsx (hunk 3 above)
Comment (LukasTy) [reply packages/x-date-pickers/src/ClockPicker/ClockPicker.tsx]: If `eslint` rules call for it, IMHO, it was a perfectly fine change.
We already default it as `false` in the function down the line.

**Review by LukasTy — APPROVED** [review]

File: packages/x-date-pickers/src/ClockPicker/ClockPicker.tsx (hunk 3 above)
Comment (Janpot) [reply packages/x-date-pickers/src/ClockPicker/ClockPicker.tsx]: I'll make a separate PR. That way we have an easy to bissect git history should it run into an unforeseen edge case

**Pull request merged**